### PR TITLE
fix cursor not available when restore state

### LINKED ISSUE #300: issue while backup and restore from savepoint
This is regarding backup restore . We are not taking any LPV backup. While restore the old LPV won't be present. We are taking savepoint backup ,while restoring getting below exception from flink connector. 
Connector should take offset from savepoint (which is backed up) and process.

Flink version used **1.9.3**, pulsar-flink-connector_2.11-**2.4.28.3** and pulsar **2.6**

**Log:**
2021-04-05 08:32:05.894 [Async calls on Source: src5 -> (src5_monitoring, invalid_data_topic_monitoring, Sink: invalid_data_topic, String_to_JsonNode -> Timestamps/Watermarks) (7/8)] WARN  o.a.flink.streaming.connectors.pulsar.internal.PulsarFetcher  - **Failed to commit cursor to Pulsar.**
java.lang.RuntimeException: **Failed to commit cursor offset 13086:38:0:0** for persistent://public/default/src5-partition-0
                    at org.apache.flink.streaming.connectors.pulsar.internal.PulsarMetadataReader.commitCursorToOffset(PulsarMetadataReader.java:263)
                    at org.apache.flink.streaming.connectors.pulsar.internal.PulsarFetcher.doCommitOffsetToPulsar(PulsarFetcher.java:452)
                    at org.apache.flink.streaming.connectors.pulsar.internal.PulsarFetcher.commitOffsetToPulsar(PulsarFetcher.java:440)
                    at org.apache.flink.streaming.connectors.pulsar.FlinkPulsarSource.notifyCheckpointComplete(FlinkPulsarSource.java:742)
                    at org.apache.flink.streaming.api.operators.AbstractUdfStreamOperator.notifyCheckpointComplete(AbstractUdfStreamOperator.java:130)
                    at org.apache.flink.streaming.runtime.tasks.StreamTask.notifyCheckpointComplete(StreamTask.java:842)
                    at org.apache.flink.runtime.taskmanager.Task$2.run(Task.java:1214)
                    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
                    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
                    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
                    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
                    at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.pulsar.client.admin.PulsarAdminException: A MultiException has 1 exceptions.  They are:
1. java.lang.IllegalStateException: ServiceLocatorImpl(__HK2_Generated_128,128,356740187) has been shut down
                    at org.apache.pulsar.client.admin.internal.BaseResource.getApiException(BaseResource.java:228)
                    at org.apache.pulsar.client.admin.internal.TopicsImpl.resetCursor(TopicsImpl.java:864)
                    at org.apache.flink.streaming.connectors.pulsar.internal.PulsarMetadataReader.commitCursorToOffset(PulsarMetadataReader.java:253)
                    ... 11 common frames omitted
Caused by: org.apache.pulsar.shade.org.glassfish.hk2.api.MultiException: A MultiException has 1 exceptions.  They are:
1. java.lang.IllegalStateException: ServiceLocatorImpl(__HK2_Generated_128,128,356740187) has been shut down
                    at org.apache.pulsar.shade.org.jvnet.hk2.internal.FactoryCreator.getFactoryHandle(FactoryCreator.java:106)
                    at org.apache.pulsar.shade.org.jvnet.hk2.internal.FactoryCreator.dispose(FactoryCreator.java:173)
                    at org.apache.pulsar.shade.org.jvnet.hk2.internal.SystemDescriptor.dispose(SystemDescriptor.java:542)
                    at org.apache.pulsar.shade.org.glassfish.jersey.inject.hk2.RequestContext.lambda$findOrCreate$0(RequestContext.java:84)
                    at org.apache.pulsar.shade.org.glassfish.jersey.internal.inject.ForeignDescriptorImpl.dispose(ForeignDescriptorImpl.java:87)
                    at org.apache.pulsar.shade.org.glassfish.jersey.inject.hk2.Hk2RequestScope$Instance.remove(Hk2RequestScope.java:150)
                    at java.lang.Iterable.forEach(Iterable.java:75)
                    at org.apache.pulsar.shade.org.glassfish.jersey.inject.hk2.Hk2RequestScope$Instance.release(Hk2RequestScope.java:167)
                    at org.apache.pulsar.shade.org.glassfish.jersey.process.internal.RequestScope.release(RequestScope.java:270)
                    at org.apache.pulsar.shade.org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:416)
                    at org.apache.pulsar.shade.org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:765)
                    at org.apache.pulsar.shade.org.glassfish.jersey.client.JerseyInvocation$Builder.method(JerseyInvocation.java:456)
                    at org.apache.pulsar.shade.org.glassfish.jersey.client.JerseyInvocation$Builder.post(JerseyInvocation.java:357)
                    at org.apache.pulsar.client.admin.internal.TopicsImpl.resetCursor(TopicsImpl.java:861)
                    ... 12 common frames omitted
Caused by: java.lang.IllegalStateException: ServiceLocatorImpl(__HK2_Generated_128,128,356740187) has been shut down
                    at org.apache.pulsar.shade.org.jvnet.hk2.internal.ServiceLocatorImpl.checkState(ServiceLocatorImpl.java:2393)
                    at org.apache.pulsar.shade.org.jvnet.hk2.internal.ServiceLocatorImpl.getServiceHandleImpl(ServiceLocatorImpl.java:635)
                    at org.apache.pulsar.shade.org.jvnet.hk2.internal.ServiceLocatorImpl.getServiceHandle(ServiceLocatorImpl.java:628)
                    at org.apache.pulsar.shade.org.jvnet.hk2.internal.ServiceLocatorImpl.getServiceHandle(ServiceLocatorImpl.java:646)
                    at org.apache.pulsar.shade.org.jvnet.hk2.internal.FactoryCreator.getFactoryHandle(FactoryCreator.java:103)
                    ... 25 common frames omitted

### PR DESCRIPTION
**Pull request by jianyun8023**
fixed #300

### CODE REVIEW

File: pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/FlinkPulsarSource.java [hunk 1]
```diff
@@ -757,6 +764,21 @@ public void initializeState(FunctionInitializationContext context) throws Except
         }
     }
 
+    private MessageId checkAndLoadAvailableCursorState(String topic, MessageId messageId) {
```
Comment (syhily) [inline pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/FlinkPulsarSource.java:767]: This logic is under discussion, we don't check the cursor status. Instead, we give user option to set their own cursor.

File: pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/FlinkPulsarSource.java [hunk 2]
```diff
@@ -467,10 +471,13 @@ public void open(Configuration parameters) throws Exception {
...
                     .forEach(e -> restoredState.put(e.getKey(), e.getValue()));
-
             restoredState.entrySet().stream()
                     .filter(e -> SourceSinkUtils.belongsTo(e.getKey(), numParallelTasks, taskIndex))
-                    .forEach(e -> ownedTopicStarts.put(e.getKey(), e.getValue()));
+                    .map(e -> {
+                        MessageId cursor = checkAndLoadAvailableCursorState(e.getKey().getTopic(), e.getValue());
+                        return Tuple2.of(e.getKey(), cursor);
+                    })
+                    .forEach(e -> ownedTopicStarts.put(e.f0, e.f1));
```
Comment (syhily) [inline pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/FlinkPulsarSource.java:480]: Use `toMap` instead?

File: pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/internal/PulsarOptions.java [hunk 3]
```diff
@@ -60,6 +60,7 @@
     public static final String ENABLE_KEY_HASH_RANGE_KEY = "enable-key-hash-range";
     public static final String KEY_DISABLED_METRICS = "key-disable-metrics";
     public static final String OLD_STATE_VERSION = "old-state-version";
+    public static final String RECOVERY_CURSOR_WHEN_CURSOR_LOST = "recovery-cursor-when-cursor-lost";
```
Comment (syhily) [inline pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/internal/PulsarOptions.java:63]: This naming is not a good choice.

File: pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/internal/SourceSinkUtils.java [hunk 4]
```diff
@@ -192,4 +194,17 @@ public static int getOldStateVersion(Map<String, String> caseInsensitiveParams,
         }
         return Integer.parseInt(value);
     }
+
+    public static MessageId getRecoveryCursorWhenCursorLost(Map<String, String> caseInsensitiveParams) {
+        final String value = caseInsensitiveParams.get(PulsarOptions.RECOVERY_CURSOR_WHEN_CURSOR_LOST);
+        if (StringUtils.isBlank(value)){
+            return null;
+        }
+        final String[] split = value.split(":");
```
Comment (syhily) [inline pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/internal/SourceSinkUtils.java:203]: Use splitter and add an extra trim?

**Review by syhily — CHANGES_REQUESTED** [review]

**Comment by jianyun8023** [comment]
Inappropriate solution, the new solution has #319 fixes.

**Pull request closed without merging**